### PR TITLE
Support Warden

### PR DESCRIPTION
**Pull request by Nerotox**
### Description of the problem being solved:
Adds support for some of the Warden nodes:
- Lesson of the Seasons
- Oath of Spring
- Oath of Summer
- Oath of Winter
- Avatar of the Wilds

Adds a generic way of handling stacking shock and scorch also easily usable by other (future) mods.

### After screenshot:
image
image
image
image
image
image
image
image
image

### CODE REVIEW

File: src/Modules/CalcPerform.lua [hunk 1]
```diff
@@ -2895,13 +2895,21 @@ function calcs.perform(env, skipEHP)
 		["Shock"] = {
 			condition = "Shocked",
 			mods = function(num)
-				return { modLib.createMod("DamageTaken", "INC", num, "Shock", { type = "Condition", var = "Shocked" }) }
+				local shockStacks = enemyDB:Sum("BASE", nil, "ShockStacks")
+				local maxShockStacks = modDB:Sum("BASE", nil, "MaxShockStacks")
+				shockStacks = maxShockStacks < 1 and 1 or shockStacks > maxShockStacks and maxShockStacks or shockStacks < 1 and 1 or shockStacks
+				output.ShockStacks = shockStacks < maxShockStacks and shockStacks or maxShockStacks
+				return { modLib.createMod("DamageTaken", "INC", num * shockStacks, "Shock", { type = "Condition", var = "Shocked" }) }
```
Comment (Paliak) [inline src/Modules/CalcPerform.lua:2902]: Should probably use a a mod property instead here.
```diff
diff --git a/src/Launch.lua b/src/Launch.lua
index 0775ec4d..694c485d 100644
--- a/src/Launch.lua
+++ b/src/Launch.lua
@@ -16,6 +16,13 @@ launch = { }
 SetMainObject(launch)
 
 function launch:OnInit()
+	-- This is the path to emmy_core.dll. The ?.dll at the end is intentional.
+	package.cpath = package.cpath .. ";Z:/home/dev/.vscode-oss/extensions/tangzx.emmylua-0.5.19/debugger/emmy/windows/x64/?.dll"
+	local dbg = require("emmy_core")
+	-- This port must match the IDE configuration. Default is 9966.
+	dbg.tcpListen("localhost", 9966)
+	-- Uncomment the next line if you want Path of Building to block until the debugger is attached
+	--dbg.waitIDE()
 	self.devMode = false
 	self.installedMode = false
 	self.versionNumber = "?"
diff --git a/src/Modules/CalcPerform.lua b/src/Modules/CalcPerform.lua
index 9955a348..0d197e8a 100644
--- a/src/Modules/CalcPerform.lua
+++ b/src/Modules/CalcPerform.lua
@@ -2895,7 +2895,7 @@ function calcs.perform(env, skipEHP)
 		["Shock"] = {
 			condition = "Shocked",
 			mods = function(num)
-				return { modLib.createMod("DamageTaken", "INC", num, "Shock", { type = "Condition", var = "Shocked" }) }
+				return { modLib.createMod("DamageTaken", "INC", num, "Shock", { type = "Condition", var = "Shocked" }, { type = "Multiplier", var = "ShockStacks", limit = modDB:Override(nil, "ShockStacksMax") or modDB:Sum("BASE", nil, "ShockStacksMax")}) }
 			end
 		},
 		["Scorch"] = {
diff --git a/src/Modules/CalcSetup.lua b/src/Modules/CalcSetup.lua
index 8954ebb4..03c1fd9d 100644
--- a/src/Modules/CalcSetup.lua
+++ b/src/Modules/CalcSetup.lua
@@ -54,6 +54,7 @@ function calcs.initModDB(env, modDB)
 	modDB:NewMod("TotemPlacementTime", "BASE", 0.6, "Base")
 	modDB:NewMod("BallistaPlacementTime", "BASE", 0.35, "Base")
 	modDB:NewMod("ActiveTotemLimit", "BASE", 1, "Base")
+	modDB:NewMod("ShockStacksMax", "BASE", 1, "Base")
 	modDB:NewMod("MovementSpeed", "INC", -30, "Base", { type = "Condition", var = "Maimed" })
 	modDB:NewMod("DamageTaken", "INC", 10, "Base", ModFlag.Attack, { type = "Condition", var = "Intimidated"})
 	modDB:NewMod("DamageTaken", "INC", 10, "Base", ModFlag.Attack, { type = "Condition", var = "Intimidated", neg = true}, { type = "Condition", var = "Party:Intimidated"})
diff --git a/src/Modules/ConfigOptions.lua b/src/Modules/ConfigOptions.lua
index c5e7033b..497359ca 100644
--- a/src/Modules/ConfigOptions.lua
+++ b/src/Modules/ConfigOptions.lua
@@ -1635,6 +1635,9 @@ Huge sets the radius to 11.
 		enemyModList:NewMod("ShockVal", "BASE", val, "Shock", { type = "Condition", var = "ShockedConfig" })
 		enemyModList:NewMod("DesiredShockVal", "BASE", val, "Shock", { type = "Condition", var = "ShockedConfig", neg = true })
 	end },
+	{ var = "multiplierShockStacks", type = "count", label = "# of ^xADAA47Shock^7 Stacks:", ifOption = "conditionEnemyShocked", tooltip = "Nubmer of stacks of ^xADAA47Shock^7 applied to the enemy", defaultState = 1, apply = function(val, modList, enemyModList)
+		enemyModList:NewMod("Multiplier:ShockStacks", "BASE", val, "Config", { type = "Condition", var = "Effective" })
+	end },
 	{ var = "conditionEnemyOnShockedGround", type = "check", label = "Is the enemy on ^xADAA47Shocked ^7Ground?", tooltip = "This also implies that the enemy is ^xADAA47Shocked.", ifEnemyCond = "OnShockedGround", apply = function(val, modList, enemyModList)
 		enemyModList:NewMod("Condition:Shocked", "FLAG", true, "Config", { type = "Condition", var = "Effective" })
 		enemyModList:NewMod("Condition:OnShockedGround", "FLAG", true, "Config", { type = "Condition", var = "Effective" })
diff --git a/src/Modules/ModParser.lua b/src/Modules/ModParser.lua
index 6dfdb6fa..f0d6a0ee 100644
--- a/src/Modules/ModParser.lua
+++ b/src/Modules/ModParser.lua
@@ -2989,6 +2989,8 @@ local specialModList = {
 	} end,
 	-- Elemental Ailments
 	["your shocks can increase damage taken by up to a maximum of (%d+)%%"] = function(num) return { mod("ShockMax", "OVERRIDE", num) } end,
+	["maximum effect of shock is (%d+)%% increased damage taken"] = function(num) return { mod("ShockMax", "OVERRIDE", num) } end,
+	["you can apply up to (%d+) shocks to each enemy"] = function(num) return { mod("ShockStacksMax", "OVERRIDE", num) } end,
 	["%+(%d+)%% to maximum effect of shock"] = function(num) return { mod("ShockMax", "BASE", num) } end,
 	["your elemental damage can shock"] = { flag("ColdCanShock"), flag("FireCanShock") },
 	["your fire damage can shock"] = { flag("FireCanShock") },

```

File: src/Modules/CalcSetup.lua [hunk 2]
```diff
@@ -54,6 +54,8 @@ function calcs.initModDB(env, modDB)
 	modDB:NewMod("TotemPlacementTime", "BASE", 0.6, "Base")
 	modDB:NewMod("BallistaPlacementTime", "BASE", 0.35, "Base")
 	modDB:NewMod("ActiveTotemLimit", "BASE", 1, "Base")
+	modDB:NewMod("ShockStacksMax", "BASE", 0, "Base")
+	modDB:NewMod("ScorchStacksMax", "BASE", 0, "Base")
```
Comment (Paliak) [inline src/Modules/CalcSetup.lua:58]: These should be 1 by defualt

Comment (Nerotox) [reply src/Modules/CalcSetup.lua:58]: Couldn't make these 1, then I'd have to reduce the num by 1 when parsing the max stack mod

Comment (Paliak) [reply src/Modules/CalcSetup.lua:58]: Ah i see. Kinda worried someone may use those without realizing that but i get it.

**Review by LocalIdentity — APPROVED** [review]

**Pull request merged**